lib/simplecov/source_file.rb

```diff
@@ -191,6 +337,205 @@ def process_skipped_lines(lines)
+    # - root branch: [:if, 0, 4, 2, 10, 5]
+    # - positive branch: [:then, 1, 5, 4, 6, 13] with root_id equal to 0
+    # - negative branch: [:else, 2, 8, 4, 9, 18] with root_id equal to 0
+    #
+    # @param [Hash] given_branches
+    #
+    # @return [Array]
+    #
+    def branches_collection(given_branches, root_id = nil)
+      @branches_collection ||= []
+
+      given_branches.each do |branch_args, value|
+        branch = Branch.new(*branch_args, root_id)
+
+        if value.is_a?(Integer)
+          branch.coverage = value
+        else
+          branches_collection(value, branch.id)
+        end
+
+        @branches_collection << branch
+      end
+      @branches_collection
+    end
+
+    #
+    # Select the covered branches.
+    #
+    # @return [Array]
+    #
+    def covered_branches
+      @coverd_branches ||= root_branches.flat_map do |root_branch|
+        root_branch.sub_branches(branches).select(&:covered?)
+      end
+    end
+
+    #
+    # Select the missed branches with coverage equal to zero
+    #
+    # @return [Array]
+    #
+    def missed_branches
+      @missed_branches ||= root_branches.flat_map do |root_branch|
+        root_branch.sub_branches(branches).select(&:missed?)
+      end
+    end
+
+    #
+    # Select the perent branches inside the branches hash
+    #
+    # @return [Array]
+    #
+    def root_branches
+      @root_branches = branches.select(&:root?)
+    end
+
+    #
+    # Method check if line is branches
+    #
+    # @param [Integer] line_number
+    #
+    # @return [Boolean]
+    #
+    def branchable_line?(line_number)
+      branches_report.keys.include?(line_number)
+    end
+
+    #
+    # Return String with branches message match to the line given
+    #
+    # @param [Integer] line_number
+    #
+    # @return [String] ex: "[1, '+'],[2, '-']" two times on negative branch and non on the positive
+    #
+    def branch_per_line(line_number)
+      branches_report[line_number].each_with_object(" ".dup) do |data, message|
+        separator = message.strip.empty? ? " " : ", "
+        message << (separator + data.to_s)
+      end.strip
+    end
+
+    #
+    # Check if any branches missing on given line number
+    #
+    # @param [Integer] line_number
+    #
+    # @return [Boolean]
+    #
+    def line_with_missed_branch?(line_number)
+      return unless branchable_line?(line_number)
+      branches_report[line_number].select { |count, _sign| count.zero? }.any?
+    end
+
+    #
+    # Build full branches report
+    # Root branches represent the wrapper of all condition state that
+    # have inside the branches
+    #
+    # @return [Hash]
+    #
+    def build_branches_report
+      root_branches.each_with_object({}) do |root_branch, statistics|
+        statistics.merge!(condition_report(root_branch))
+      end
+    end
+
+    #
+    # Create hash as branches coverage report
+    # keys: lines numbers matching the branch start line
```

Comment:
I feel like most of the terminology should use singular not plural. I don't believe it will cause confusion, and it reads easier. So: "branch coverage" instead of "branches coverage", "key: line number" instead of "key: lines numbers" etc.